### ISSUE TITLE
mass-deployer: testing checklist

### ISSUE THREAD

**Issue by xmonader** [threefoldtech/tfgrid-sdk-go#710]
## User Interaction Test Cases
- [x] User-Friendly Interface: Evaluate the user-friendliness of the command-line interface.
- [x] Comprehensive Documentation: Test the availability and clarity of documentation for users.
- [x] Error Guidance: Assess the system’s guidance for resolving common user errors.
- [x] Feedback Mechanism for Users: Evaluate the effectiveness of the feedback mechanism for user inputs.
- [x] Deployment Progress Feedback: Check for real-time progress feedback during deployment.
- [x] Cancellation of Deployment: Test the cancellation of deployments.
- [x] Detailed Error Messages: Ensure error messages are detailed and informative.
- [x] First-Time User Deployment Experience: Evaluate the experience of a first-time user deploying on a node group.
- [x] Guided Deployment Process: Test the effectiveness of a guided deployment process for new users.
- [x] Compatibility with Different Operating Systems: Test tfrobot on various operating systems.
- [x] Invalid Configuration File: Test behavior with a malformed or invalid configuration file.
- [x] Deployment with Insufficient Resources: Attempt to deploy with insufficient resources in a node group.
- [x] Network Failure During Deployment: Simulate a network failure during deployment and observe recovery mechanisms.
- [x] Incorrect SSH Key: Deploy with an incorrect SSH key and observe error handling? Can we validate SSH keys?
- [x] Failure in VM Creation: Test system behavior when VM creation fails in a node group.
- [x] Invalid Region Specification: Test behavior with an invalid or non-existent region specification.
- [x] Zero VMs in Configuration: Test behavior when the configuration specifies zero VMs.
- [x] Exceeding Maximum Resource Limits: Attempt to deploy with resource specifications above the system's maximum limits.
- [x] Deployment with Minimum Resource Threshold: Deploy using the minimum possible resource specifications.

## Functionality Test Cases

- [x] Single Node Group Deployment: Deploy on single node group with basic configuration.
- [x] Multiple Node Groups Deployment: Deploy on multiple node groups simultaneously.
- [x] Multiple VMs Deployment in a Node Group: Deploy multiple VMs in a single node group.
- [x] VM Deployment with Specific CPU and Memory: Test VM deployment with specific CPU and memory requirements.
- [x] Node Group with Public IPv4: Deploy a node group with public IPv4 enabled.
- [x] Node Group with Public IPv6: Deploy a node group with public IPv6 enabled.
- [x] VM Deployment with SSD and HDD Storage: Deploy VMs with specific SSD and HDD configurations.
- [x] Complex Node Group Configurations: Test deployment with complex node group configurations.
- [x] Environmental Variables Setup: Test the setting up of environmental variables for VMs.
- [x] Custom SSH Key Management: Explore the usage of different SSH keys for different VMs.
- [x] Being able to SSH to VMs using flists that support SSH

## Performance Test Cases
- [x] Large Scale Deployment: Deploy a large number of VMs across multiple node groups.
- [x] #836
- [x] Resource Utilization During Deployment: Monitor CPU, memory, and network utilization during deployment.

## Reliability Test Cases
- [x] Automatic Retry on Failure: Evaluate the effectiveness of the automatic retry mechanism.
- [x] Recovery from Partial Deployments: Test the system's ability to recover from partially completed deployments??

### Test Case: Basic Deployment YAML
- Command: `tfrobot deploy -c /configs/config.yaml`
- Description: Deploy VMs using the basic configuration file.

### Test Case: Basic Deployment yml
- Command: `tfrobot deploy -c /configs/config.yaml`
- Description: Deploy VMs using the basic configuration file.
### Test Case: Basic Deployment JSON
- Command: `tfrobot deploy -c /configs/config.json`
- Description: Deploy VMs using the basic configuration file.

### Test Case: Deployment with Output Specification
- Command: `tfrobot deploy -c /configs/config.yaml -o /output/out.yaml`
- Description: Deploy VMs using configuration file and save the output information to a specified file.
### Test Case: Deployment with Debug Mode
- Command: `tfrobot deploy -c /configs/config.yaml -d`
- Description: Deploy VMs with a configuration designed for debugging, enabling debug logs in the output.
### Test Case: Canceling Deployment
- Command: `tfrobot cancel -c /configs/config.yaml`
- Description: Cancel all VMs deployed using the basic configuration file.
- 
### Test Case: Canceling Deployment with Output Specification
- Command: `tfrobot cancel -c /configs/config.yaml`
- Description: Cancel all VMs deployed using configuration file and save the output information to a specified file.
### Test Case: Canceling Deployment with Debug Mode
- Command: `tfrobot cancel -c /configs/config.yaml -d`
- Description: Cancel deployments with a configuration designed for debugging, enabling debug logs in the output.
### Test Case: Help for Deploy
- Command: `tfrobot deploy -h`
- Description: Display help information for the deploy subcommand.
### Test Case: Help for Cancel
- Command: `tfrobot cancel -h`
- Description: Display help information for the cancel subcommand.
### Test Case: Help for the Main Command
- Command: `tfrobot -h`
- Description: Display general help information for the tfrobot tool.
### Test Case: Deployment with All Flags
- Command: `tfrobot deploy -c /configs/config.yaml -o /output/out.yaml -d`
- Description: Deploy VMs using a complex configuration file, enabling debug logs, and saving the output to a specific file.

### Test Case: Missing Configuration File Path
- Command: `tfrobot deploy`
- Expected Error: "Error: Missing required '-c' flag for configuration file path."

### Test Case: Unreadable/Nonexistent Configuration File Path
- Command: `tfrobot deploy -c /unreadable/path/config.yaml` 
- Expected Error: "Error: Missing required '-c' flag for configuration file path."

### Test Case: Invalid Configuration File Path
- Command: `tfrobot deploy -c /invalid/path/config.yaml`
- Expected Error: "Error: Configuration file not found at specified path."
### Test Case: Missing Subcommand
- Command: `tfrobot -c /configs/config.yaml`
- Expected Error: "Error: Subcommand required (deploy or cancel)."
### Test Case: Unsupported Flags
- Command: `tfrobot deploy -x`
- Expected Error: "Error: Unsupported flag '-x'. Use '-h' for help."
### Test Case: Deploy Without Configuration File
- Command: `tfrobot deploy -o /output/out.yaml`
- Expected Error: "Error: '-c' flag with configuration file path is required."
### Test Case: Cancel Without Configuration File
- Command: `tfrobot cancel -o /output/out.yaml`
- Expected Error: "Error: '-c' flag with configuration file path is required."
### Test Case: Multiple Configuration Files
- Command: `tfrobot deploy -c /configs/config1.yaml -c /configs/config2.yaml`
- Expected Error: "Error: Multiple configuration files specified. Please use only one '-c' flag."

### Test Case: Deploy with Non-writable Output Path
- Command: `tfrobot deploy -c /configs/config.yaml -o /nonwritable/output/path/info.yaml`
- Expected Error: "Error: Output path does not exist or is not writable."

### Test Case: Deploy with Output path that's not .json, .yaml or .yml
- Command: `tfrobot deploy -c /configs/basic_config.yaml -o /nonwritable/output/path/info.txt`
- Expected Error: "Error: Output path needs to end with .json or .yaml."

### Test Case: Use of Deploy and Cancel Together
- Command: ` tfrobot deploy cancel -c /configs/config.yaml`
- Expected Error: "Error: 'deploy' and 'cancel' cannot be used together. Please use one command at a time."
### Test Case : Configuration File with Invalid Format
- Command: `tfrobot deploy -c /configs/config.yaml`
- Expected Error: "Error: Configuration file format is invalid. Please check the file and try again."
### Test Case: Using -d Flag Without Subcommand
- Command: `tfrobot -d -c /configs/config.yaml`
- Expected Error: "Error: Subcommand required when using '-d'. Use '-h' for help."
### Test Case: Help Flag with Subcommand
- Command: `tfrobot deploy -h -c /configs/config.yaml``
- Expected Error: "Warning: '-h' flag used with other commands. Ignoring other flags and showing help for 'deploy'."

**Comment by A-Harby** [comment]
### Work Completed:
I created a TFRobot test suite: https://app.testlodge.com/a/26076/projects/40893/suites/236972 With 4 sections: docs & installation, input validation, mass deployer, and mass canceler, I proceeded with filling out the two sections of docs & installation and input validation.
And started testing on documents and installation, and is still working on the input validation section.
Created issue:
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/717
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/719
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/720
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/721
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/722
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/723
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/724

**Comment by A-Harby** [comment]
Created issue:
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/730
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/731
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/733
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/734
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/735
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/738

**Comment by rawdaGastan** [comment]
## User Interaction Test Cases
- [ ] User-Friendly Interface: Evaluate the user-friendliness of the command-line interface.
- [ ] Comprehensive Documentation: Test the availability and clarity of documentation for users.
- [ ] Error Guidance: Assess the system’s guidance for resolving common user errors.
- [ ] Feedback Mechanism for Users: Evaluate the effectiveness of the feedback mechanism for user inputs.
- [x] Deployment Progress Feedback: Check for real-time progress feedback during deployment.
- [x] Cancellation of Deployment: Test the cancellation of deployments.
- [ ] Detailed Error Messages: Ensure error messages are detailed and informative.
- [ ] First-Time User Deployment Experience: Evaluate the experience of a first-time user deploying on a node group.
- [ ] Guided Deployment Process: Test the effectiveness of a guided deployment process for new users.
- [ ] Compatibility with Different Operating Systems: Test tfrobot on various operating systems.
- [ ] Invalid Configuration File: Test behavior with a malformed or invalid configuration file.
- [x] Deployment with Insufficient Resources: Attempt to deploy with insufficient resources in a node group.
- [ ] Network Failure During Deployment: Simulate a network failure during deployment and observe recovery mechanisms.
- [x] Incorrect SSH Key: Deploy with an incorrect SSH key and observe error handling? Can we validate SSH keys?
- [x] Failure in VM Creation: Test system behavior when VM creation fails in a node group.
- [x] Invalid Region Specification: Test behavior with an invalid or non-existent region specification.
- [x] Zero VMs in Configuration: Test behavior when the configuration specifies zero VMs.
- [x] Exceeding Maximum Resource Limits: Attempt to deploy with resource specifications above the system's maximum limits.
- [x] Deployment with Minimum Resource Threshold: Deploy using the minimum possible resource specifications.

## Functionality Test Cases

- [x] Single Node Group Deployment: Deploy on single node group with basic configuration.
- [x] Multiple Node Groups Deployment: Deploy on multiple node groups simultaneously.
- [x] Multiple VMs Deployment in a Node Group: Deploy multiple VMs in a single node group.
- [ ] VM Deployment with Specific CPU and Memory: Test VM deployment with specific CPU and memory requirements.
- [x] Node Group with Public IPv4: Deploy a node group with public IPv4 enabled.
- [x] Node Group with Public IPv6: Deploy a node group with public IPv6 enabled.
- [ ] VM Deployment with SSD and HDD Storage: Deploy VMs with specific SSD and HDD configurations.
- [ ] Complex Node Group Configurations: Test deployment with complex node group configurations.
- [ ] Environmental Variables Setup: Test the setting up of environmental variables for VMs.
- [ ] Custom SSH Key Management: Explore the usage of different SSH keys for different VMs.
- [x] Being able to SSH to VMs using flists that support SSH

## Performance Test Cases
- [ ] Large Scale Deployment: Deploy numerous VMs across multiple node groups.
- [ ] Deployment Speed Test: Measure the time taken for deploying VMs on a node group and VMs on multiple node groups.
- [ ] Resource Utilization During Deployment: Monitor CPU, memory, and network utilization during deployment.

## Reliability Test Cases
- [x] Automatic Retry on Failure: Evaluate the effectiveness of the automatic retry mechanism.
- [ ] Recovery from Partial Deployments: Test the system's ability to recover from partially completed deployments??

## Command line arguments interaction

- test cases with outputs: [command line arguments interaction](https://github.com/threefoldtech/tfgrid-sdk-go/files/14125733/cmd_args_tests.md)

**Comment by xmonader** [comment]
![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/e9e9679c-d181-4505-af8b-b99dc99775d2)
![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/cc6d95e7-7d40-45ae-bd69-8db1d3d29741)
![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/f980c1d3-e730-4656-80b2-fb33ecd793b3)
![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/ae56b70b-fd50-4186-bff1-d014063011ea)
![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/6868d03a-cfe0-48f6-a94d-9a0ed4afa4fd)
![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/74150111-cb06-4eba-8b6d-d3cddb4a6816) this one was a nice touch to enable planetary in case of no pub ip4 or 6


![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/26244586-953d-4230-8664-7b9b898d6378)

![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/8516d03a-ba0a-464c-9618-7bd3d53fab28)

![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/1f5ce30b-456c-47c1-bd2b-7f23c9a2947d)

![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/fc371e71-7ae3-40b4-9b77-64d7489b77b0)

![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/84f2f723-6c46-43e3-ae2b-577d386923f3)

![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/bd2df7ba-48a4-4b30-bb8a-e6def4de2371)

![image](https://github.com/threefoldtech/tfgrid-sdk-go/assets/64129/5cf3b5dc-2419-497a-ba31-1e46b1b3c798)

**Comment by A-Harby** [comment]
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/742
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/744
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/745
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/746

**Comment by A-Harby** [comment]
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/756
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/757
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/768

**Comment by A-Harby** [comment]
Work completed:
- Created a full test suite to cover mass deployer cases: https://app.testlodge.com/a/26076/projects/40893/suites/236972.
- All the issues found were reported, and I reviewed all the closed issues to add test cases for them in the test suite.

**Comment by A-Harby** [comment]
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/794
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/792
- https://github.com/threefoldtech/tfgrid-sdk-go/issues/791

**Comment by A-Harby** [comment]
> VM Deployment with SSD and HDD Storage: Deploy VMs with specific SSD and HDD configurations.

There isn't a way now to deploy HDD storage, and as per the comment [here](https://github.com/threefoldtech/tfgrid-sdk-go/issues/791), it will be applied when ZDB is added.
 
> Network Failure During Deployment: Simulate a network failure during deployment and observe recovery mechanisms.

I'm waiting for the [issue](https://github.com/threefoldtech/tfgrid-sdk-go/issues/745) to be done, to re-verify it, and to mark the point as done. 

>  Recovery from Partial Deployments: Test the system's ability to recover from partially completed deployments??

I couldn't see anything in the documents about this point, and the current approach I found is that when mass deployment fails, other than the part that got deployed, the other part won't be deployed again.

> Network Failure During Deployment: Simulate a network failure during deployment and observe recovery mechanisms.

I tried a deployment of 1 node group with 1 VM group of 10 VMs, and it took 44 seconds. I then tried the same deployment, but in the middle, I disconnected the network for 5 minutes and waited.
The VMs were successful, but it took 16 minutes to finish the deployment (or 11m without 5m of internet disconnect).

> Incorrect SSH Key: Deploy with an incorrect SSH key and observe error handling? Can we validate SSH keys?

It has a ssh key validation if the format is incorrect ssh key `example1 is invalid: ssh: no key found`.

I will also create an issue and update it with Deployment Speed Test and Resource Utilization During Deployment.
 - https://github.com/threefoldtech/tfgrid-sdk-go/issues/836
